source/common/network/apple_dns_impl.cc

```diff
@@ -52,60 +52,69 @@ AppleDnsResolverStats AppleDnsResolverImpl::generateAppleDnsResolverStats(Stats:
   return {ALL_APPLE_DNS_RESOLVER_STATS(POOL_COUNTER(scope))};
 }
 
+AppleDnsResolverImpl::StartResolutionResult
+AppleDnsResolverImpl::startResolution(const std::string& dns_name,
+                                      DnsLookupFamily dns_lookup_family, ResolveCb callback) {
+  auto pending_resolution =
+      std::make_unique<PendingResolution>(*this, callback, dispatcher_, dns_name);
+
+  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+  if (error != kDNSServiceErr_NoError) {
+    ENVOY_LOG(warn, "DNS resolver error ({}) in dnsServiceGetAddrInfo for {}", error, dns_name);
+    chargeGetAddrInfoErrorStats(error);
+    return {nullptr, false};
+  }
+
+  if (pending_resolution->synchronously_completed_) {
+    return {nullptr, true};
+  }
+
+  // Hook up the query's UDS socket to the event loop to process updates.
+  if (!pending_resolution->dnsServiceRefSockFD()) {
+    ENVOY_LOG(warn, "DNS resolver error in dnsServiceRefSockFD for {}", dns_name);
+    return {nullptr, false};
+  }
+
+  return {std::move(pending_resolution), true};
+}
+
 ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
```

Comment:
Let me give that a go

Reply:
Ok now all the logic is inside the helper and error handling happens outside of the helper. I considered also returning the address in case of inline resolution, but that would require digging into the PendingResolution class to get that out somehow, which seemed a bit high risk